src/Nethermind/Nethermind.Synchronization/ParallelSync/MultiSyncModeSelector.cs

```diff
@@ -136,84 +127,90 @@ public void Update()
-                                CheckAddFlag(best.IsInFastSync, SyncMode.FastSync, ref newModes);
-                                CheckAddFlag(best.IsInFullSync, SyncMode.Full, ref newModes);
-                                CheckAddFlag(best.IsInStateSync && !canBeInSnapRangesPhase, SyncMode.StateNodes,
-                                    ref newModes);
-                                CheckAddFlag(best.IsInStateSync && canBeInSnapRangesPhase, SyncMode.SnapSync,
-                                    ref newModes);
-                                CheckAddFlag(best.IsInDisconnected, SyncMode.Disconnected, ref newModes);
-                                CheckAddFlag(best.IsInWaitingForBlock, SyncMode.WaitingForBlock, ref newModes);
-                                if (IsTheModeSwitchWorthMentioning(newModes))
-                                {
-                                    string stateString = BuildStateString(best);
-                                    if (_logger.IsInfo)
-                                        _logger.Info($"Changing state {Current} to {newModes} at {stateString}");
-                                }
+                                newModes = SyncMode.Full;
                             }
-                            catch (InvalidAsynchronousStateException)
+                            else
                             {
                                 newModes = SyncMode.Disconnected;
-                                reason = "Snapshot Misalignment";
+                                reason = "No Useful Peers";
                             }
                         }
-                        
-
-
-                        if ((newModes & (SyncMode.Full | SyncMode.WaitingForBlock)) != SyncMode.None
-                            && (Current & (SyncMode.Full | SyncMode.WaitingForBlock)) == SyncMode.None)
+                    }
+                    else
+                    {
+                        try
+                        {
+                            best.IsInFastSync = ShouldBeInFastSyncMode(best);
+                            best.IsInStateSync = ShouldBeInStateSyncMode(best);
+                            best.IsInFullSync = ShouldBeInFullSyncMode(best);
+                            best.IsInFastHeaders = ShouldBeInFastHeadersMode(best);
+                            best.IsInFastBodies = ShouldBeInFastBodiesMode(best);
+                            best.IsInFastReceipts = ShouldBeInFastReceiptsMode(best);
+                            best.IsInDisconnected = ShouldBeInDisconnectedMode(best);
+                            best.IsInWaitingForBlock = ShouldBeInWaitingForBlockMode(best);
+                            bool canBeInSnapRangesPhase = CanBeInSnapRangesPhase(best);
+
+                            newModes = SyncMode.None;
+                            CheckAddFlag(best.IsInBeaconHeaders, SyncMode.BeaconHeaders, ref newModes);
```

Comment:
Where in fastSync we're checking if we should be in beaconHeaders or BeaconModeControl?